File: src/main/java/macro303/neptunes/display/models/TeamsModel.java
package macro303.neptunes.display.models;

import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import macro303.neptunes.Team;
import macro303.neptunes.game.Game;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Comparator;

/**
 * Created by Macro303 on 2018-05-08.
 */
public class TeamsModel implements Model {
	@NotNull
	private final ObservableList<Team> teams = FXCollections.observableList(new ArrayList<>());

	@Override
	public void updateModel(@NotNull Game game) {
		teams.setAll(game.getTeams());
		teams.sort(Comparator.reverseOrder());
	}

	@NotNull
	public ObservableList<Team> getTeams() {
		return teams;
	}
}

File: src/main/kotlin/macro303/neptunes/team/TeamsModel.kt
package macro303.neptunes.team

import javafx.beans.property.SimpleBooleanProperty
import javafx.collections.FXCollections
import macro303.neptunes.Model
import macro303.neptunes.game.Game
import tornadofx.*
import java.time.format.DateTimeFormatter
import java.util.concurrent.Executor
import java.util.concurrent.Executors

/**
 * Created by Macro303 on 2018-05-08.
 */
class TeamsModel : ViewModel(), Model {
	val teams = FXCollections.observableList(ArrayList<Team>())
	var game: Game? = null
		private set
	override val loading = SimpleBooleanProperty(false)
	private val exec: Executor

	init {
		exec = Executors.newCachedThreadPool { runnable ->
			val t = Thread(runnable)
			t.isDaemon = true
			t
		}
		updateModel()
	}

	@Suppress("UNCHECKED_CAST")
	override fun updateModel() {
		val refreshTask = TeamRefreshTask(this)
		refreshTask.setOnFailed {
			println("Unable to connect to Team API: ${refreshTask.exception.message}")
			loading.value = false
		}
		refreshTask.setOnSucceeded {
			this.teams.clear()
			val results = it.source.value as ArrayList<Team>?
			if (results != null) {
				this.teams.setAll(results)
				this.teams.sortDescending()
			}
			println("Team API loaded")
			loading.value = false
		}
		refreshTask.setOnCancelled {
			println("Unable to connect to Team API")
			loading.value = false
		}
		exec.execute(refreshTask)
	}

	override fun getAddress(): String {
		return "players"
	}

	companion object {
		private val dateFormat = DateTimeFormatter.ofPattern("MMM-dd HH:mm")
		private val timeFormat = DateTimeFormatter.ofPattern("MMM-dd HH:mm:ss")
	}
}